shared/health/src/main/java/io/pravega/shared/health/impl/ContributorRegistryImpl.java

```diff
@@ -0,0 +1,321 @@
+     * will be registered when no parent {@link HealthComponent} is specified.
+     */
+    private final HealthComponent root;
+
+    private final String rootName;
+
+    @NonNull
+    public ContributorRegistryImpl(String rootName, StatusAggregator aggregator) {
+        this.rootName = rootName;
+        this.root = new HealthComponent(rootName, aggregator, this);
+        initialize(root);
+    }
+
+    @NonNull
+    public ContributorRegistryImpl(String rootName) {
+        this(rootName, StatusAggregatorImpl.DEFAULT);
+    }
+
+    private void initialize(HealthComponent root) {
+        // Initialize children set of the root.
+        children.put(rootName, new HashSet<>());
+        // Not needed, but useful for validation.
+        parents.put(rootName, new HashSet<>());
+        // Make it visible to clients.
+        contributors.put(rootName, root);
+        // Separates 'contributors' from 'components' -- useful for an internal vs leaf node distinction.
+        components.add(rootName);
+    }
+
+    /**
+     * Registers a {@link HealthComponent} under the implicit 'root' {@link HealthComponent}.
+     *
+     * @param component The {@link HealthComponent} object to add to the registry.
+     * @return The {@link HealthContributor} registered under the {@link ContributorRegistry}.
+     */
+    @NonNull
+    @Override
+    synchronized public HealthContributor register(HealthComponent component) {
+        return register(component, root);
+    }
+
+    /**
+     * Registers a {@link HealthContributor} under the implicit 'root' {@link HealthComponent}.
+     *
+     * @param contributor The {@link HealthContributor} object to add to the registry.
+     * @return The {@link HealthContributor} registered under the {@link ContributorRegistry}.
+     */
+    @NonNull
+    @Override
+    synchronized public HealthContributor register(HealthContributor contributor) {
+        return register(contributor, root);
+    }
+
+    /**
+     * Registers a {@link HealthComponent} under the specified parent {@link HealthComponent}.
+     *
+     * @param component The {@link HealthComponent} object to add to the registry as a child of 'parent'.
+     * @param parent The {@link HealthComponent} to add 'contributor' as a child/dependency.
+     * @return The {@link HealthContributor} registered under the {@link ContributorRegistry}.
+     */
+    @NonNull
+    @Override
+    synchronized public HealthContributor register(HealthComponent component, HealthComponent parent) {
+        parent = Objects.isNull(parent) ? root : parent;
+        // A HealthComponent should only exist if defined during construction, instead of adding it dynamically.
+        if (!components.contains(parent.getName()) || !contributors.containsKey(parent.getName())) {
+            log.warn("Attempting to register {} under unrecognized {} -- aborting.", component, parent);
```

Comment:
Is this log warn necessary? I think throwing an exception would be better.

Would the caller expect a `null` as a return value?

Reply:
No, an exception would be better. No calls actually use the `HealthContributor` return value at the moment, so I'm considering just making this method `void`.